internal/k8s/deviceplugin/server.go

```diff
@@ -0,0 +1,308 @@
+	logEntry := log.Entry(ctx).WithField("devicePluginServer", "StartServers")
+
+	logEntry.Info("get resource lister client")
+	resourceListerClient, err := manager.GetPodResourcesListerClient(ctx)
+	if err != nil {
+		logEntry.Error("failed to get resource lister client")
+		return err
+	}
+
+	resp, err := resourceListerClient.List(ctx, new(podresources.ListPodResourcesRequest))
+	if err != nil {
+		logEntry.Errorf("resourceListerClient unavailable: %+v", err)
+		return err
+	}
+	_ = tokenPool.Restore(respToDeviceIDs(resp))
+
+	for name := range tokenPool.Tokens() {
+		s := &devicePluginServer{
+			ctx:                  ctx,
+			name:                 name,
+			tokenPool:            tokenPool,
+			resourcePollTimeout:  resourcePollTimeout,
+			updateCh:             make(chan struct{}, 1),
+			allocatedTokens:      map[string]bool{},
+			resourceListerClient: resourceListerClient,
+		}
+
+		tokenPool.AddListener(s.update)
+
+		logEntry.Infof("starting server: %v", name)
+		socket, err := manager.StartDeviceServer(s.ctx, s)
+		if err != nil {
+			logEntry.Errorf("error starting server: %v", name)
+			return err
+		}
+
+		logEntry.Infof("registering server: %s", name)
+		if err := manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+			Version:      pluginapi.Version,
+			Endpoint:     socket,
+			ResourceName: name,
+		}); err != nil {
+			logEntry.Errorf("error registering server: %s", name)
+			return err
+		}
+
+		if err := s.monitorKubeletRestart(manager, socket); err != nil {
+			logEntry.Warnf("error monitoring kubelet restart: %s %+v", name, err)
+		}
+	}
+	return nil
+}
+
+func respToDeviceIDs(resp *podresources.ListPodResourcesResponse) map[string][]string {
+	deviceIDs := map[string][]string{}
+	for _, pod := range resp.PodResources {
+		for _, container := range pod.Containers {
+			for _, device := range container.Devices {
+				deviceIDs[device.ResourceName] = append(deviceIDs[device.ResourceName], device.DeviceIds...)
+			}
+		}
+	}
+	return deviceIDs
+}
+
+func (s *devicePluginServer) update() {
+	select {
+	case s.updateCh <- struct{}{}:
+	default:
+	}
+}
+
+func (s *devicePluginServer) monitorKubeletRestart(manager K8sManager, socket string) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "monitorKubeletRestart")
+
+	resetCh, err := manager.MonitorKubeletRestart(s.ctx)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		logEntry.Infof("start monitoring kubelet restart: %s", s.name)
+		defer logEntry.Infof("stop monitoring kubelet restart: %s", s.name)
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case _, ok := <-resetCh:
+				if !ok {
+					return
+				}
+				logEntry.Infof("re registering server: %s", s.name)
+				if err = manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+					Version:      pluginapi.Version,
+					Endpoint:     socket,
+					ResourceName: s.name,
+				}); err != nil {
+					logEntry.Errorf("error re registering server: %s %+v", s.name, err)
```

Comment:
Should we do a fatal here? Since it will not continue working properly and will not recover.

Reply:
Yep, fixed.